libraries/api-extractor-model/src/mixins/ApiExtendsMixin.ts

```diff
@@ -0,0 +1,316 @@
+}
+
+export interface IApiExtendsMixinJson extends IApiItemJson {
+  extendsTokenRanges: IExcerptTokenRange[];
+}
+
+const _extendsTypes: unique symbol = Symbol('ApiExtendsMixin._extendsTypes');
+
+/**
+ * The mixin base class for API items that can extend other API items.
+ *
+ * @remarks
+ *
+ * This is part of the {@link ApiModel} hierarchy of classes, which are serializable representations of
+ * API declarations.  The non-abstract classes (e.g. `ApiClass`, `ApiEnum`, `ApiInterface`, etc.) use
+ * TypeScript "mixin" functions (e.g. `ApiDeclaredItem`, `ApiItemContainerMixin`, etc.) to add various
+ * features that cannot be represented as a normal inheritance chain (since TypeScript does not allow a child class
+ * to extend more than one base class).  The "mixin" is a TypeScript merged declaration with three components:
+ * the function that generates a subclass, an interface that describes the members of the subclass, and
+ * a namespace containing static members of the class.
+ *
+ * @public
+ */
+// eslint-disable-next-line @typescript-eslint/naming-convention
+export interface ApiExtendsMixin extends ApiItem {
+  /**
+   * The types referenced via an "extends" heritage clause.
+   */
+  readonly extendsTypes: HeritageType[];
+
+  /**
+   * Resolves all of the ApiItem's immediate and inherited members by walking down the
+   * inheritance tree.
+   */
+  resolveInheritedMembers(): IResolveInheritedMembersResult;
+
+  /** @override */
+  serializeInto(jsonObject: Partial<IApiExtendsMixinJson>): void;
+}
+
+/**
+ * Mixin function for {@link (ApiExtendsMixin:interface)}.
+ *
+ * @param baseClass - The base class to be extended
+ * @returns A child class that extends baseClass, adding the {@link (ApiExtendsMixin:interface)} functionality.
+ *
+ * @public
+ */
+export function ApiExtendsMixin<TBaseClass extends IApiItemConstructor>(
+  baseClass: TBaseClass
+  // eslint-disable-next-line @typescript-eslint/no-explicit-any
+): TBaseClass & (new (...args: any[]) => ApiExtendsMixin) {
+  class MixedClass extends baseClass implements ApiExtendsMixin {
+    public readonly [_extendsTypes]: HeritageType[] = [];
+
+    // eslint-disable-next-line @typescript-eslint/no-explicit-any
+    public constructor(...args: any[]) {
+      super(...args);
+
+      const options: IApiExtendsMixinOptions = args[0];
+
+      if (this instanceof ApiDeclaredItem) {
+        for (const extendsTokenRange of options.extendsTokenRanges) {
+          this.extendsTypes.push(new HeritageType(this.buildExcerpt(extendsTokenRange)));
+        }
+      } else {
+        throw new InternalError('ApiExtendsMixin expects a base class that inherits from ApiDeclaredItem');
+      }
+    }
+
+    /**
+     * Resolves all of the ApiItem's immediate and inherited members by walking down the
+     * inheritance tree. Be aware that his method returns both immediate and inherited members,
+     * which might not be evident from its name.
+     *
```

Comment:
> might not be evident from its name

Maybe we can improve the name?  The "resolving" word seems like an implementation detail, not the main idea of this API.  A word like "find" or "search" might help convey that nontrivial computational work is happening, and that we might not find every item.  (`findMembersWithInheritance()`?)

🤔 On the other hand, it seems like found inherited members are importantly different from the "actual" direct members.  For example, https://github.com/microsoft/rushstack/issues/3429#issuecomment-1133819235 suggested a presentation like:

![image](https://user-images.githubusercontent.com/4673363/169679603-f757ec80-3512-481e-bcd5-87d9e9c353db.png)

So maybe it would actually be preferable for `findInheritedMembers()` to return ONLY the inherited members that do coincide with actual members, and then it is the documenter's job to mix the two sets together?

Reply:
For a bit more context, I chose "resolve" to match the `resolveDeclarationReference` method. The two methods were similar in that they both are resolving references under the hood and return similar shaped result objects. I don't feel strongly though, so I'm fine with `resolveMembersWithInheritance` or `findMembersWithInheritance`.

> So maybe it would actually be preferable for findInheritedMembers() to return ONLY the inherited members that do coincide with actual members, and then it is the documenter's job to mix the two sets together?

The downside to be aware of here is that the documenter will need to sort the combined sets. Currently, the documenter doesn't have to do any sorting on its own.

Reply:
> The downside to be aware of here is that the documenter will need to sort the combined sets. Currently, the documenter doesn't have to do any sorting on its own.

😅

Reply:
🤔 I think I prefer `findMembersWithInheritance()` to `resolveInheritedMembers()`. I think of "resolving" as trying to locate the thing that an identifier is reference. That is part of this operation, but it seems like a less significant part than crawling a graph and searching for things, possibly with some query parameters, and possibly failing to find some things (that might be found by a better implementation).  In some of the more esoteric cases that we discussed, such as deducing the "members" from type algebra, `resolveDeclarationReference()` may not be involved at all.

The `WithInheritance` suffix was supposed to clarify that we find both inherited and non-inherited members, whereas `resolveInheritedMembers()` sounded like it excludes non-inherited members.

Reply:
(Looks like C# has both [`GetMembers()`](https://docs.microsoft.com/en-us/dotnet/api/system.type.getmembers?view=net-6.0#system-type-getmembers) and [`FindMembers()`](https://docs.microsoft.com/en-us/dotnet/api/system.type.findmembers?view=net-6.0#system-type-findmembers(system-reflection-membertypes-system-reflection-bindingflags-system-reflection-memberfilter-system-object)) -- maybe that influenced my thinking somehow heheh.)

Reply:
Done.